upgrades/upgrade_system.gd

```diff
@@ -1,30 +1,29 @@
 class_name UpgradeSystem
 extends Resource
 
-var _upgrades: Array
+var _upgrades: Array[Upgrade]
 
-var _state: GameState
+var _purse: CoinPurse
 
-func _init(state) -> void:
+func _init(purse: CoinPurse) -> void:
 	_upgrades = []
-	_state = state
+	_purse = purse
```

Comment:
everything works and makes sense for how purchases work, but i'm wondering we could just pass the purse into the `_handle_attempt_upgrade` signal call instead of needing it in the constructor.

or if we should have the upgrade system even know about the purse - don't need to change anything, but just thinking

Reply:
I attempted to pass in the get_gold function instead, but I couldn't find a way to properly add the static type checking for the function, so opted for this instead.